File: crypto_bot/src/data/cleaner.py
"""Sort, dedupe, clamp OHLC, forward-fill a single missing bar, flag large gaps."""

from __future__ import annotations

import pandas as pd

INTERVAL = {"15m": "15min", "1h": "1h", "4h": "4h", "1d": "1d"}


def clean_ohlcv(df: pd.DataFrame, timeframe: str) -> pd.DataFrame:
    if df.empty:
        return df.copy()
    out = df.copy()
    if not isinstance(out.index, pd.DatetimeIndex):
        raise ValueError("OHLCV index must be DatetimeIndex UTC")
    if out.index.tz is None:
        out.index = out.index.tz_localize("UTC")
    else:
        out.index = out.index.tz_convert("UTC")
    out = out.sort_index()
    out = out[~out.index.duplicated(keep="last")]
    for col in ("open", "high", "low", "close", "volume"):
        if col not in out.columns:
            raise ValueError(f"missing column {col}")
        out[col] = pd.to_numeric(out[col], errors="coerce")
    out = out.dropna(subset=["open", "high", "low", "close"])
    out["high"] = out[["high", "open", "close"]].max(axis=1)
    out["low"] = out[["low", "open", "close"]].min(axis=1)
    out["volume"] = out["volume"].clip(lower=0)
    freq = INTERVAL[timeframe]
    full = pd.date_range(out.index.min(), out.index.max(), freq=freq, tz="UTC")
    out = out.reindex(full)
    gap_bars = out["close"].isna()
    # forward-fill a single missing bar from previous close
    out[["open", "high", "low", "close"]] = out[["open", "high", "low", "close"]].ffill(limit=1)
    out.loc[out["volume"].isna() & out["close"].notna(), "volume"] = 0
    out["gap"] = gap_bars & out["close"].notna()
    still = out["close"].isna()
    out.loc[still, "gap"] = True
    out = out.dropna(subset=["close"])
    return out

File: src/lib/quant/data/cleaner.ts
import { INTERVAL_MS } from "../config";
import type { Ohlcv, Timeframe } from "../types";

const EPS = 1e-12;

function isValidBar(bar: Ohlcv): boolean {
  return (
    Number.isFinite(bar.open) &&
    Number.isFinite(bar.high) &&
    Number.isFinite(bar.low) &&
    Number.isFinite(bar.close) &&
    Number.isFinite(bar.volume) &&
    bar.high + EPS >= Math.max(bar.open, bar.close, bar.low) &&
    bar.low - EPS <= Math.min(bar.open, bar.close, bar.high) &&
    bar.volume >= 0
  );
}

/**
 * Sort by time, drop duplicates, clamp OHLC, forward-fill a single missing bar,
 * and flag gaps larger than 1.5 intervals.
 */
export function cleanOhlcv(raw: Ohlcv[], timeframe: Timeframe): Ohlcv[] {
  const interval = INTERVAL_MS[timeframe];
  const byTime = new Map<number, Ohlcv>();
  for (const bar of raw) {
    if (!isValidBar(bar)) continue;
    const t = Math.floor(bar.time / interval) * interval;
    byTime.set(t, {
      time: t,
      open: bar.open,
      high: Math.max(bar.high, bar.open, bar.close, bar.low),
      low: Math.min(bar.low, bar.open, bar.close, bar.high),
      close: bar.close,
      volume: bar.volume,
    });
  }

  const times = [...byTime.keys()].sort((a, b) => a - b);
  if (times.length === 0) return [];

  const out: Ohlcv[] = [];
  for (let i = 0; i < times.length; i++) {
    const t = times[i]!;
    const bar = byTime.get(t)!;
    if (i > 0) {
      const prev = out[out.length - 1]!;
      const dt = t - prev.time;
      if (dt > interval * 1.5 && dt <= interval * 2.5) {
        out.push({
          time: prev.time + interval,
          open: prev.close,
          high: prev.close,
          low: prev.close,
          close: prev.close,
          volume: 0,
          gap: true,
        });
      } else if (dt > interval * 2.5) {
        bar.gap = true;
      }
    }
    out.push(bar);
  }
  return out;
}
